tests/utils.py

```diff
@@ -0,0 +1,110 @@
+"""
+Utility functions for pywbem testing
+"""
+
+from __future__ import absolute_import
+
+import sys
+import os
+from packaging import version
+import pytest
+import ply
+
+
+def skip_if_moftab_regenerated():
+    """
+    Skip the testcase if we run against an installed version of pywbem (as
+    indicated by the TEST_INSTALLED environment variable), and the testcase
+    would re-generate the MOF parsing table files.
+
+    This function should be used by test cases that parse MOF files, so the
+    test case is skipped in that case.
+
+    Background:
+
+    Pywbem uses the `ply` package to parse CIM MOF files and part of a pywbem
+    installation are parsing table files named mofparsetab.py and moflextab.py
+    that are generated by `ply`. Before pywbem 0.14.5, the version of `ply`
+    module that was used to build the pywbem distribution archive needed to be
+    compatible with the version of `ply` installed in the Python environment,
+    otherwise `ply` attempted to re-generate these parse table files in the
+    `pywbem` installation directory. Thus, if the `pywbem` installation
+    directory is in the system Python, a normal user will typically encounter
+    a permission denied error.
+
+    If the installed version of `pywbem` is 0.14.5 or higher, it has tolerance
+    against mismatches between these `ply` versions, by having `ply`
+    re-generate the parsing tables in memory if needed, but no longer writing
+    them out to the pywbem installation directory.
+    """
+
+    pywbem = import_installed('pywbem')  # noqa: E402
+    from pywbem import mofparsetab, moflextab
+
+    test_installed = os.getenv('TEST_INSTALLED', False)
+
+    pywbem_not_tolerant = version.parse(pywbem.__version__) <= \
+        version.parse('0.14.4')  # This causes 0.14.5.devN to be tolerant
+    mofparsetab_mismatch = version.parse(mofparsetab._tabversion) != \
+        version.parse(ply.__version__)
+    moflextab_mismatch = version.parse(moflextab._tabversion) != \
+        version.parse(ply.__version__)
+
+    if test_installed and pywbem_not_tolerant and \
+            (mofparsetab_mismatch or moflextab_mismatch):
+        pytest.skip("The installed pywbem version {0} does not tolerate ply "
+                    "version mismatches, and there is a ply version mismatch: "
+                    "ply: {1}, mofparsetab: {2}, moflextab: {3}".
+                    format(pywbem.__version__, ply.__version__,
+                           mofparsetab._tabversion, moflextab._tabversion))
+
+
+def import_installed(module_name):
+    """
+    Import a Python module/package, controlling whether it is loaded from the
+    normal Python module search path, or from an installed version (excluding
+    the module in the current directory).
+
+    The TEST_INSTALLED environment variable controls this as follows:
+
+      * If not set (or set to an empty string), the normal Python module search
```

Comment:
Why include the option of setting to an empty string with the not set.  Isn't it easier to simply use
existences or non-existence of the env var??

Reply:
DONE - After first thinking it is strange that it should be enabled when set to an empty string, I looked at other variables in the makefile and they all enable their function when just being defined, regardless of the value. That was also the case for TEST_INSTALLED in the makefile, so it was even inconsistently implemented. So I agree with the suggestion and have changed the code accordingly.